source/adapters/cuda/enqueue.cpp

```diff
@@ -484,6 +484,133 @@ UR_APIEXPORT ur_result_t UR_APICALL urEnqueueCooperativeKernelLaunchExp(
                                numEventsInWaitList, phEventWaitList, phEvent);
 }
 
+UR_APIEXPORT ur_result_t UR_APICALL urEnqueueKernelLaunchCustomExp(
+    ur_queue_handle_t hQueue, ur_kernel_handle_t hKernel, uint32_t workDim,
+    const size_t *pGlobalWorkSize, const size_t *pLocalWorkSize,
+    uint32_t numAttrsInLaunchAttrList,
+    const ur_exp_launch_attribute_t *launchAttrList,
+    uint32_t numEventsInWaitList, const ur_event_handle_t *phEventWaitList,
+    ur_event_handle_t *phEvent) {
+
```

Comment:
In general you need to make sure there is the necessary migration of buffer and image mem between devices in a context. It would suffice to assert that the kernel has no `ur_mem_handle_t` args.

Reply:
See here https://github.com/oneapi-src/unified-runtime/blob/main/source/adapters/cuda/enqueue.cpp#L421 as well as other places in `urEnqueueKernelLaunch`

Reply:
Thanks, I forgot about the multi-context patch. I've now made this consistent with it.

Reply:
Nice one